File: desktop/Netsira.Desktop/Diagnostics/DiagnosticRunService.cs
// SPDX-License-Identifier: AGPL-3.0-or-later
using Netsira.Desktop.Devices;

namespace Netsira.Desktop.Diagnostics;

public enum DiagnosticMode
{
    Quick,
    Standard,
    Comprehensive
}

public sealed record AirOsConnectionSettings(
    string Host,
    string Username,
    string Password,
    bool AllowInvalidCertificate);

public sealed record DiagnosticRunResult(
    DiagnosticMode Mode,
    QuickDiagnosticResult Quick,
    Ndt7Result? Speed,
    AirOsSnapshot? AirOs,
    IReadOnlyList<string> Errors,
    IReadOnlyList<string> Skipped);

public sealed class DiagnosticRunService
{
    private readonly QuickDiagnosticService _quick = new();
    private readonly MlabNdt7Client _ndt7 = new();

    public async Task<DiagnosticRunResult> RunAsync(
        DiagnosticMode mode,
        AirOsConnectionSettings? airOsSettings,
        CancellationToken cancellationToken = default)
    {
        var errors = new List<string>();
        var skipped = new List<string>();
        var quick = await _quick.RunAsync(cancellationToken);

        AirOsSnapshot? airOs = null;
        Ndt7Result? speed = null;

        if (mode == DiagnosticMode.Comprehensive)
        {
            if (airOsSettings is null ||
                string.IsNullOrWhiteSpace(airOsSettings.Host) ||
                string.IsNullOrWhiteSpace(airOsSettings.Username) ||
                string.IsNullOrEmpty(airOsSettings.Password))
            {
                skipped.Add("CPE status: device address or credentials were not fully provided.");
            }
            else
            {
                try
                {
                    using var client = new AirOsClient(airOsSettings.AllowInvalidCertificate);
                    airOs = await client.ConnectAndReadAsync(
                        airOsSettings.Host,
                        airOsSettings.Username,
                        airOsSettings.Password,
                        cancellationToken);
                }
                catch (Exception ex)
                {
                    errors.Add("CPE status failed: " + ex.Message);
                }
            }
        }

        if (mode != DiagnosticMode.Quick)
        {
            if (!quick.HasNetwork)
            {
                skipped.Add("M-Lab throughput: no active network was detected.");
            }
            else
            {
                try
                {
                    speed = await _ndt7.RunAsync(cancellationToken);
                }
                catch (Exception ex)
                {
                    errors.Add("M-Lab throughput failed: " + ex.Message);
                }
            }
        }

        return new DiagnosticRunResult(mode, quick, speed, airOs, errors, skipped);
    }
}

File: android/app/src/main/java/io/github/shadowur0/netsira/diagnostics/DiagnosticRunService.kt
// SPDX-License-Identifier: AGPL-3.0-or-later
package io.github.shadowur0.netsira.diagnostics

import android.content.Context
import io.github.shadowur0.netsira.devices.AirOsClient
import io.github.shadowur0.netsira.devices.AirOsSnapshot
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

enum class DiagnosticMode {
    Quick,
    Standard,
    Comprehensive
}

data class AirOsConnectionSettings(
    val host: String,
    val username: String,
    val password: String
)

data class DiagnosticRunResult(
    val mode: DiagnosticMode,
    val quick: QuickDiagnosticResult,
    val speed: Ndt7Result?,
    val airOs: AirOsSnapshot?,
    val errors: List<String>,
    val skipped: List<String>
)

class DiagnosticRunService(private val context: Context) {
    suspend fun run(
        mode: DiagnosticMode,
        airOsSettings: AirOsConnectionSettings?
    ): DiagnosticRunResult {
        val errors = mutableListOf<String>()
        val skipped = mutableListOf<String>()
        val quick = QuickDiagnostic(context).run()

        var airOs: AirOsSnapshot? = null
        var speed: Ndt7Result? = null

        if (mode == DiagnosticMode.Comprehensive) {
            if (airOsSettings == null ||
                airOsSettings.host.isBlank() ||
                airOsSettings.username.isBlank() ||
                airOsSettings.password.isEmpty()
            ) {
                skipped += "CPE status: device address or credentials were not fully provided."
            } else {
                try {
                    airOs = withContext(Dispatchers.IO) {
                        AirOsClient().connectAndRead(
                            airOsSettings.host,
                            airOsSettings.username,
                            airOsSettings.password
                        )
                    }
                } catch (e: Exception) {
                    errors += "CPE status failed: " + (e.message ?: e.javaClass.simpleName)
                }
            }
        }

        if (mode != DiagnosticMode.Quick) {
            if (!quick.hasNetwork) {
                skipped += "M-Lab throughput: no active network was detected."
            } else {
                try {
                    speed = MlabNdt7Client().run()
                } catch (e: Exception) {
                    errors += "M-Lab throughput failed: " + (e.message ?: e.javaClass.simpleName)
                }
            }
        }

        return DiagnosticRunResult(mode, quick, speed, airOs, errors, skipped)
    }
}
